File: sdk/python/aegisrun/step.py
"""Step execution management"""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any, Dict, Optional

from ulid import ULID

from .client import AegisRunClient
from .tool_call import ToolCall, ToolCallBlockedError
from .types import StepStatus

if TYPE_CHECKING:
    from .run import Run

logger = logging.getLogger(__name__)


class Step:
    """Represents a single step within a run"""

    def __init__(
        self,
        client: AegisRunClient,
        run_id: str,
        seq_no: int,
        name: str,
        state_vector: Dict[str, Any],
        run: Optional["Run"] = None,
    ):
        self.client = client
        self.run_id = run_id
        self.run = run
        self.seq_no = seq_no
        self.name = name
        self.state_vector = state_vector

        self.step_id = str(ULID())
        self.status = StepStatus.RUNNING
        self.started_at: Optional[datetime] = None
        self.ended_at: Optional[datetime] = None

        self._tool_call_seq = 0

    def start(self):
        """Mark step as started"""
        self.started_at = datetime.now(timezone.utc)
        self._emit_event(
            event_type="step.started",
            payload={
                "step_id": self.step_id,
                "name": self.name,
                "seq_no": self.seq_no,
            },
            timestamp=self.started_at,
        )

    def complete(self):
        """Mark step as completed"""
        self.status = StepStatus.COMPLETED
        self.ended_at = datetime.now(timezone.utc)
        self._emit_event(
            event_type="step.ended",
            payload={
                "step_id": self.step_id,
                "status": "completed",
                "seq_no": self.seq_no,
            },
            timestamp=self.ended_at,
        )

    def fail(self, error: str):
        """Mark step as failed"""
        self.status = StepStatus.FAILED
        self.ended_at = datetime.now(timezone.utc)
        self._emit_event(
            event_type="step.ended",
            payload={
                "step_id": self.step_id,
                "status": "failed",
                "seq_no": self.seq_no,
                "error": error,
            },
            timestamp=self.ended_at,
        )

    def _emit_event(
        self,
        event_type: str,
        payload: Dict[str, Any],
        timestamp: Optional[datetime] = None,
    ) -> None:
        try:
            ts = (timestamp or datetime.now(timezone.utc)).isoformat()
            self.client.submit_event(
                run_id=self.run_id,
                event_type=event_type,
                payload=payload,
                timestamp=ts,
            )
        except Exception:
            logger.warning(
                "Failed to emit %s for run=%s step=%s",
                event_type,
                self.run_id,
                self.step_id,
                exc_info=True,
            )

    def tool_call(
        self,
        tool_name: str,
        args: Dict[str, Any],
        executor: str = "builtin",
    ) -> Any:
        """Execute a tool call through the gateway"""
        tool_call = ToolCall(
            client=self.client,
            run_id=self.run_id,
            step_id=self.step_id,
            seq_no=self._tool_call_seq,
            tool_name=tool_name,
        )

        self._tool_call_seq += 1

        try:
            result = tool_call.execute(
                args=args,
                state_vector=self.state_vector,
                executor=executor,
            )
            if self.run:
                self.run.counters.tool_calls += 1
            return result
        except ToolCallBlockedError:
            if self.run:
                self.run.counters.blocks += 1
            raise

File: sdk/typescript/src/step.ts
/**
 * Step execution management
 */

import { ulid } from 'ulid';
import { AegisRunClient } from './client';
import { ToolCall, ToolCallBlockedError } from './toolCall';
import { StepStatus, RunCounters } from './types';

export interface StepConfig {
  client: AegisRunClient;
  runId: string;
  seqNo: number;
  name: string;
  stateVector: Record<string, any>;
  counters?: RunCounters;
}

export class Step {
  private client: AegisRunClient;
  private runId: string;
  private seqNo: number;
  private name: string;
  private stateVector: Record<string, any>;

  public stepId: string;
  public status: StepStatus = StepStatus.Running;
  public startedAt?: Date;
  public endedAt?: Date;

  private toolCallSeq = 0;
  private counters?: RunCounters;

  constructor(config: StepConfig) {
    this.client = config.client;
    this.runId = config.runId;
    this.seqNo = config.seqNo;
    this.name = config.name;
    this.stateVector = config.stateVector;
    this.counters = config.counters;

    this.stepId = ulid();
  }

  start(): void {
    this.startedAt = new Date();
    this.emitEvent('step.started', {
      step_id: this.stepId,
      name: this.name,
      seq_no: this.seqNo,
    }, this.startedAt);
  }

  complete(): void {
    this.status = StepStatus.Completed;
    this.endedAt = new Date();
    this.emitEvent('step.ended', {
      step_id: this.stepId,
      status: 'completed',
      seq_no: this.seqNo,
    }, this.endedAt);
  }

  fail(error: string): void {
    this.status = StepStatus.Failed;
    this.endedAt = new Date();
    this.emitEvent('step.ended', {
      step_id: this.stepId,
      status: 'failed',
      seq_no: this.seqNo,
      error,
    }, this.endedAt);
  }

  private emitEvent(
    eventType: string,
    payload: Record<string, any>,
    timestamp: Date
  ): void {
    const maybePromise = this.client.submitEvent({
      runId: this.runId,
      eventType,
      payload,
      timestamp: timestamp.toISOString(),
    });

    Promise.resolve(maybePromise).catch(() => {
      // Best-effort lifecycle telemetry; do not fail step execution.
    });
  }

  async toolCall(
    toolName: string,
    args: Record<string, any>,
    executor: string = 'builtin'
  ): Promise<any> {
    const toolCall = new ToolCall({
      client: this.client,
      runId: this.runId,
      stepId: this.stepId,
      seqNo: this.toolCallSeq,
      toolName,
    });

    this.toolCallSeq++;

    try {
      const result = await toolCall.execute({
        args,
        stateVector: this.stateVector,
        executor,
      });
      if (this.counters) {
        this.counters.tool_calls++;
      }
      return result;
    } catch (error) {
      if (error instanceof ToolCallBlockedError && this.counters) {
        this.counters.blocks++;
      }
      throw error;
    }
  }
}
